src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -172,13 +176,296 @@ struct TensorContract
+  }
+
+  // Inserts the first contracted TensorIndex into the list of contracted LHS
+  // TensorIndexs
+  template <typename... LhsIndices>
+  using get_uncontracted_lhs_tensorindex_list_helper = tmpl::append<
+      tmpl::at_c<tmpl::split_at<tmpl::list<LhsIndices...>,
+                                tmpl::size_t<FirstContractedIndexPos>>,
+                 0>,
+      tmpl::list<tmpl::at_c<ArgsList, FirstContractedIndexPos>>,
+      tmpl::at_c<tmpl::split_at<tmpl::list<LhsIndices...>,
+                                tmpl::size_t<FirstContractedIndexPos>>,
+                 1>>;
+
+  /// Constructs the uncontracted LHS's list of TensorIndexs by inserting the
+  /// pair of indices being contracted into the list of contracted LHS
+  /// TensorIndexs
+  ///
+  /// Example: If we contract RHS tensor \f${R^{a}}_{bac}\f$ to LHS tensor
+  /// \f$L_{cb}\f$, the RHS list of generic indices (`ArgsList`) is
+  /// `tmpl::list<ti_A_t, ti_b_t, ti_a_t, ti_c_t>` and the LHS generic indices
+  /// (`LhsIndices`) are `ti_c_t, ti_b_t`. `ti_A_t` and `ti_a_t` are inserted
+  /// into `LhsIndices` at their positions from the RHS, which yields:
+  /// `tmpl::list<ti_A_t, ti_c_t, ti_a_t, ti_b_t>`.
+  template <typename... LhsIndices>
+  using get_uncontracted_lhs_tensorindex_list = tmpl::append<
+      tmpl::at_c<tmpl::split_at<get_uncontracted_lhs_tensorindex_list_helper<
+                                    LhsIndices...>,
+                                tmpl::size_t<SecondContractedIndexPos>>,
+                 0>,
+      tmpl::list<tmpl::at_c<ArgsList, SecondContractedIndexPos>>,
+      tmpl::at_c<tmpl::split_at<get_uncontracted_lhs_tensorindex_list_helper<
+                                    LhsIndices...>,
+                                tmpl::size_t<SecondContractedIndexPos>>,
+                 1>>;
+
+  /// \brief Helper struct for computing the contraction of one pair of indices
+  ///
+  /// \tparam UncontractedLhsStructure the Structure of the uncontracted LHS
+  /// tensor
+  /// \tparam UncontractedLhsTensorIndexList the typelist of TensorIndexs of the
+  /// uncontracted LHS tensor
+  /// \tparam ContractedLhsNumComponents the number of components in the
+  /// contracted LHS tensor
+  /// \tparam T1 the expression type contained within the RHS contraction
+  /// expression
+  /// \tparam Index for a given list of uncontracted LHS storage indices whose
+  /// components are summed to compute a contracted LHS component, this is the
+  /// position of one such storage index in that list
+  template <typename UncontractedLhsStructure,
+            typename UncontractedLhsTensorIndexList,
+            size_t ContractedLhsNumComponents, typename T1, size_t Index>
+  struct ComputeContraction;
+
+  template <typename UncontractedLhsStructure,
+            typename... UncontractedLhsTensorIndices,
+            size_t ContractedLhsNumComponents, typename T1, size_t Index>
+  struct ComputeContraction<UncontractedLhsStructure,
```

Comment:
I'd recommend putting `Index`, `ContractedLhsNumComponents`, `T1`, and `ContractedLhsNumComponents`as template parameters on the `apply` function to avoid creating more types/classes